amaranth/vendor/_gatemate.py

```diff
@@ -0,0 +1,120 @@
+from abc import abstractmethod
+from amaranth import *
+from amaranth.build import *
+from amaranth.lib.cdc import ResetSynchronizer
+
+__all__ = ["GateMatePlatform"]
+
+class GateMatePlatform(TemplatedPlatform):
+    """
+    .. rubric:: GateMate FPGA toolchain
+
+    Required tools:
+        * ``yosys``
+        * ``p_r``
+
+    The environment is populated by setting the ``CC_TOOL`` environment variable to point to the toolchain directory.
+    """
+
+    device = property(abstractmethod(lambda: None))
+    package = property(abstractmethod(lambda: None))
+
+    toolchain = "GateMate"
+
+    required_tools = [
+        "yosys",
+        "p_r",
+    ]
+
+    file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.ccf": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                Net "{{port_name}}" Loc = "{{pin_name}}"
+                {%- set constraints = [
+                    "SCHMITT_TRIGGER", "PULLUP", "PULLDOWN", "KEEPER", "SLEW", "DRIVE", 
+                    "DELAY_OBF", "DELAY_IBF", "FF_IBF", "FF_OBF", "LVDS_BOOST", "LVDS_TERM"
+                ] -%}
+                {%- for constraint in constraints if constraint in attrs -%}
+                    | {{constraint}}={{attrs[constraint]}}
+                {%- endfor -%};
+            {% endfor %}
+        """,
+        "{{name}}.sdc": r"""
+            # {{autogenerated}}
+            {% for net_signal, port_signal, frequency in platform.iter_clock_constraints() -%}
+                {% if port_signal is not none -%}
+                    create_clock -period {{1000000000/frequency}} [get_ports {{port_signal.name|tcl_quote}}]
+                {% endif %}
+            {% endfor %}
+        """,
+    }
+
+    command_templates = [
+        r"""
+        mkdir -p log;
+        mkdir -p net;
+        {{invoke_tool("yosys")}}
+            -ql log/synth.log -p 'read -sv {{name}}.v; synth_gatemate -top {{name}} -nomx8 -vlog net/{{name}}_synth.v'
```

Comment:
The Yosys script should be another generated file.

`-q` should only be given if requested by the end user.

You can copy the Yosys invocation from the SiliconBlue platform.

Reply:
I have tried this now, but I get some errors that the top module is missing. I will try to resolve tomorrow.